### PR TITLE
PEN-1374 Shared Infra ALB

### PR DESCRIPTION
**Pull request by bil-sukhtsooj**
# Summary

Create CDK stack that contains shared ALB to be used by our platform app target groups.

# AWS 
We are using L2 construct for creating the ALBs to our specific needs for this shared infra.

Currently we have following resources this stack:

1. ALB (for external apps)
- Two default listeners(`80, 443`) and third listener for port `8081`.
- Security group
     - Outbound: allow all
     - Inbound: allow all
2. ALB (for internal apps) 
- Default two listeners for ports `80, 443`
- Security group
    - Default allow all outbound rule
    - Inbound rule to only allow traffic from ips in the range `10.0.0.0/8`

### CODE REVIEW

File: cdk/package.json [hunk 1]
```diff
@@ -10,11 +10,11 @@
         "build": "tsc",
         "cdk": "cdk synthesize",
         "cdk:destroy:proto": "NODE_ENV=proto cdk destroy --all --profile whcc-proto",
-        "cdk:proto": "NODE_ENV=proto cdk deploy infrastructure-application-stack-proto-ue2  --profile whcc-proto",
-        "cdk:proto:prereq": "NODE_ENV=proto cdk deploy infrastructure-prereqstack-proto-ue2 --profile whcc-proto",
-        "cdk:destroy:dev": "NODE_ENV=dev AWS_PROFILE=whcc-fe-dev cdk destroy --all",
-        "cdk:dev": "NODE_ENV=dev AWS_PROFILE=whcc-fe-dev cdk deploy infrastructure-application-stack-dev-ue2",
-        "cdk:dev:prereq": "NODE_ENV=dev AWS_PROFILE=whcc-fe-dev cdk deploy infrastructure-prereqstack-dev-ue2"
+        "cdk:proto": "NODE_ENV=proto cdk deploy platform-shared-infra-proto-ue2  --profile whcc-proto",
+        "cdk:proto:prereq": "NODE_ENV=proto cdk deploy platform-prereqstack-proto-ue2 --profile whcc-proto",
```
Comment (PatrickNiesen) [inline cdk/package.json:14]: Small change I had both **--profile** and **AWS_PROFILE** in here for testing. I would standardize them to one or the other

**Review by PatrickNiesen — CHANGES_REQUESTED** [review]
Looks great, and I think this will be a good thing to build/update going forward too

File: cdk/package.json (hunk 1 above)
Comment (bil-sukhtsooj) [reply cdk/package.json]: Standardized all of them to use `AWS_PROFILE`

File: cdk/lib/constructs/albl2.ts [hunk 2]
```diff
@@ -1,58 +1,77 @@
...
-    certArns: string[],
-    albExtraListeners: alb.BaseApplicationListenerProps[],
-    albInternetFacing: boolean,
-  }
-
+export interface AlbL2Props extends cdk.StackProps {
+  albName: string,
+  vpc: ec2.IVpc,
+  certArns: string[],
+  albExtraListeners?: alb.BaseApplicationListenerProps[],
```
[the remark this answers is not in the document]
Comment (bil-sukhtsooj) [reply cdk/lib/constructs/albl2.ts:11]: I agree to renaming with prefix `alb` for consistency.

File: cdk/bin/cdk.ts [hunk 3]
```diff
@@ -57,13 +57,12 @@ const prereqStack = new PrereqStack(cdkApp, {
...
-  ...awsConfig, 
+  ...awsConfig,
   albName: namingConfig.GetAwsNaming("alb"),
   vpcName: namingConfig.GetAwsNaming("vpc"),
-  ecrName: namingConfig.GetAwsNaming("ecr"),
-  ecrImageTag: 'latest',
   certArns: awsConfig.resources.application.certArns,
-  stackName: namingConfig.GetAwsNaming("application-stack")
+  stackName: namingConfig.GetAwsNaming("shared-infra"),
+  environment: process.env.NODE_ENV!
```
Comment (timturnerwhcc) [inline cdk/bin/cdk.ts:65]: Help me understand why we need this passed all the way through?

Comment (troydwhcc) [reply cdk/bin/cdk.ts:65]: There are several layers, and each layer is responsible for part of the naming. We suspect that in the future we may need to support overrides at the top level.
- Naming Convention: `[<global-identifier> "-" ] <component-name> "-" <optional-detail> "-" <aws-identifier> "-" <environment>[ "-" <region>]`
- Component Stack: The stack is responsible for the `component-name`.
- L2 Construct/Module: Owns the `optional-detail`, which is generally only applicable if we have multiple of the same resource type in the same stack.
- L1 Construct/Module: Owns the `was-identifier`. This is the `naming` `resourceEnum`.
- `environment` and `region` should come from the environment configuration (JSON config).

Comment (timturnerwhcc) [reply cdk/bin/cdk.ts:65]: This makes sense.

My specific questionwas on `environment: process.env.NODE_ENV!` We don't seem to actually _use_ the `environment` on the `props` in the Stack CDK itself (unless I missed it). The `environment` coming from the `NODE_ENV` defined in `package.json` I could be convinced is a good idea, but it doesn't feel strict enough?

Comment (bil-sukhtsooj) [reply cdk/bin/cdk.ts]: @timturnerwhcc Good catch! This is unused, I forgot to remove it. I was using `NODE_ENV` in the stack before for condition to create external alb. But then refactored it to use flag from config Json.

File: cdk/bin/cdk.ts [hunk 4]
```diff
@@ -57,13 +57,11 @@ const prereqStack = new PrereqStack(cdkApp, {
 });
 
 const applicationStack = new ApplicationStack(cdkApp,namingConfig,awsResoucesConfig,{
-  ...awsConfig, 
+  ...awsConfig,
   albName: namingConfig.GetAwsNaming("alb"),
```
[the remark this answers is not in the document]
Comment (troydwhcc) [reply cdk/bin/cdk.ts:61]: I assume that the L3 constructs would not expose this.

**Comment by bil-sukhtsooj** [comment]
Addressed all the changes we've discussed last Friday.

File: cdk/bin/cdk.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (PatrickNiesen) [reply cdk/bin/cdk.ts:53]: I would think we would want L3 to be the construct that imports the AWSNaming Libraries. Is that what were you thinking? Users of the L3 constructs would need minimum inputs towards naming the components.

[the remark this answers is not in the document]
Comment (troydwhcc) [reply cdk/bin/cdk.ts:53]: I agree with that.

**Review by PatrickNiesen — APPROVED** [review]

File: cdk/config/default.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
   "aws": {
     "env": {
-      "account": "689368240727",
+      "account": "799497006720",
```
Comment (andrew-merkouris-whcc) [inline cdk/config/default.json:4]: @bil-sukhtsooj @PatrickNiesen Can we remove default properties which will not get used?

(Same question for all the below properties which are blank)

File: cdk/config/dev.json [hunk 6]
```diff
@@ -12,24 +12,27 @@
...
+          "certArns": [ "arn:aws:acm:us-east-2:799497006720:certificate/ce6306c4-27ee-4d93-b12a-fafab73552af" ],
+          "createInternalAlb": true,
+          "createExternalAlb": false
         }
       }
     },
     "naming": {
       "globalidentifier": "whcc",
-      "componentname": "infrastructure",
+      "componentname": "platform",
```
Comment (andrew-merkouris-whcc) [inline cdk/config/dev.json:35]: @bil-sukhtsooj @PatrickNiesen Should we **not** override properties where not necessary?

**Review by timturnerwhcc — APPROVED** [review]
This addresses the concerns I had. I don't feel my approval is sufficient in isolation for merging.

File: cdk/config/default.json (hunk 5 above)
Comment (bil-sukhtsooj) [reply cdk/config/default.json]: Yes. I cleaned up all unused properties and config file.

**Review by andrew-merkouris-whcc — APPROVED** [review]

**Pull request merged**